### PR TITLE
execution/stagedsync: pre-write tx nonces to drop sender-based scheduling in parallel exec

### PR DESCRIPTION
**Pull request by yperbasis**
Fixes erigontech/metarepo-internal-issues#19.

## What

On the no-BAL parallel-execution path, every tx with a same-sender predecessor used to be `clearPending`'d out of the initial wave (sender-based dependency scheduling). The only state that actually *requires* this serialization is the sender nonce — and the nonce is declared by the tx itself, so it is statically known for any valid block.

`processRequest` now pre-writes each tx's output nonce (`tx.Nonce + 1`, plain `uint64`, matching what `SetNonce`/BAL pre-population store at `NoncePath`) into the block's `VersionMap` at the tx's own index, and schedules everything optimistically — same as the BAL branch. A same-sender successor's `GetNonce` (preCheck) reads its predecessor's pre-written output via the normal MVCC floor read instead of failing with nonce-too-high and waiting on a scheduler dependency.

Type check requested by the issue ("Verify before coding"): the `NoncePath` versioned value is a plain `uint64` everywhere (`versionWritten` from `SetNonce`, BAL `WriteChanges`, `valuesEqual`, `versionedUpdate[uint64]`), so the pre-write carries `uint64(tx.Nonce()+1)` — byte/type-exact.

Kept as-is per the issue:
- the `Dependencies()` branch (Polygon/Bor; `USE_TX_DEPENDENCIES`),
- the BAL branch (`len(accessList) != 0`) — BALs already pre-populate nonces.

Excluded from pre-writes (these keep the legacy sender-dependency path):
- AA txs (RIP-7560 nonce semantics differ — `NonceManager`),
- `tx.Nonce == MaxUint64` (`+1` would wrap; such a tx invalidates the block anyway, so don't inject a wrapped value into the shared map).

Kill switch: `EXEC3_NONCE_PREWRITES` (default `true`); `false` restores sender-based scheduling — handy for devnet A/B comparison (including `IGNORE_BAL=true` runs on BAL devnets, which take exactly this branch).

## Deviation from the issue: step 3 ("validate rather than apply") is not needed

The issue proposed making the nonce increment at `txn_executor.go:535` (and the EIP-7702 authority bump) validate-only in pre-write mode, to guard against a double increment. After tracing the MVCC read/write paths I left the state transition untouched, because the double increment is structurally impossible and skipping the write would actually break things:

- **No self-read:** `VersionMap.Read(addr, path, key, txIdx)` floors at `txIdx-1` — a tx can never observe its *own* pre-written entry. `GetNonce` returns the predecessor's value, so `SetNonce(nonce+1)` writes exactly the pre-written value: the flush is an idempotent same-value overwrite at the same `(TxIndex, Incarnation=0)` in the common case.
- **The three-way assert already exists:** `mapvalue == txvalue` is preCheck's nonce check (`GetNonce` reads the map under versioned execution), and `read == post-predecessor state` is Block-STM read validation. A divergence either rejects the tx (genuinely invalid nonce) or re-executes it via the normal path — exactly the issue's intended semantics.
- **Skipping the write would break read-your-own-writes:** the EIP-7702 self-authorization check reads the authority nonce *after* the tx-level bump (`verifyAuthorities`, step 5); without the local versioned write it would read the stale predecessor value from the map.
- **Skipping the write would break EIP-7928 BAL generation:** `AsBlockAccessList` builds nonce changes from the per-tx write sets (`blockIO` outputs); removing the nonce from `TxOut` would drop nonce changes from generated BALs.
- **Self-correcting divergence:** for a valid self-authorization (authority == sender) the real output nonce is `tx.Nonce + 2`; the execution write overwrites the pre-written `tx.Nonce + 1` at flush time, so successors re-validate against the correct value.
- **The pre-written cell can never be deleted:** every successful incarnation re-emits the sender nonce write (so the re-execution `prevWrites` cleanup never deletes the cell), failed incarnations never record writes, the sender is always dirty in its own tx (so `MakeWriteSet`'s reverted-address `DeleteAll` never fires for it), and block-end finalize uses a separate local version map.

## Expected effect

Same-sender txs now run fully in the initial parallel wave (the IO-warming speculative pass from the issue/profiling discussion). They typically still re-execute once on the genuine sender-*balance* (gas debit) conflict after their predecessor flushes — as the issue states, that is an ordinary per-path version-map conflict, identical to any two txs touching the same account, and the re-run hits warm state.

## Tests (TDD)

Red → green unit tests driving the scheduler change (`execution/stagedsync/exec3_nonce_prewrite_test.go`, run `processRequest` directly and inspect the dispatch wave + version map):
- `TestProcessRequestPreWritesNonces` — same-sender txs all dispatched in the initial wave; `NoncePath` cells at each tx's own index with value `nonce+1`, incarnation 0; floor read returns the predecessor's pre-write (red on main: 3/5 dispatched, no cells).
- `TestProcessRequestPreWritesNoncesSetCodeTx` — type-4 txs pre-write `tx.Nonce+1` regardless of authorizations (red on main).
- `TestProcessRequestNoPreWriteForAATxs`, `TestProcessRequestNoPreWriteForMaxNonce`, `TestProcessRequestSenderDepsWithPreWritesDisabled` — exclusions and kill switch keep legacy behavior.

End-to-end acceptance tests (`execution/tests/parallel_exec_same_sender_test.go`); these are regression guards per the issue's acceptance list — correctness was already preserved by the old serialization, so they pass before and after, but now exercise the optimistic path:
- `TestSameSenderTxsInOneBlock` — one block with 5 same-sender txs (incl. a contract creation, which bumps the nonce inside `evm.Create`) plus another sender; inserted under the parallel executor (`WithExperimentalBAL` forces it) and the serial executor; asserts final nonces (no double increment: sender ends at exactly 5, created contract at 1) and that the two executors produce identical account records; the state root is cross-checked by `InsertChain` against the serially-generated header.
- `TestSameSenderNonceGapRejected` — a hand-built block with a same-sender nonce gap is rejected by both executors (the gapped tx fails preCheck against the predecessor's pre-written nonce).

EIP-7702 authority-nonce edge cases (incl. authority == a later in-block sender) are covered by the eest fixtures that CI runs under the `EXEC3_PARALLEL=true` leg; building a Prague `GenerateChain` harness for an in-repo e2e was out of proportion for this PR.

## Verification

- Instrumented `processRequest` with a temporary probe to confirm the e2e test really engages the parallel executor and the new branch (6 probe hits, one per tx) before removing it.
- `go test ./execution/stagedsync ./execution/state/... ./execution/exec/...` green.
- `go test -race ./execution/stagedsync` green; `EXEC3_PARALLEL=true go test -race ./execution/tests -run 'TestSameSender|TestSelfDestructReceive|TestInsertChain'` green.
- `EXEC3_PARALLEL=true go test ./execution/tests` (full parallel leg) green.
- `EXEC3_NONCE_PREWRITES=false go test ./execution/stagedsync` green (kill-switch leg).
- `make lint` clean; `make erigon integration` build.

## Out of scope (per the issue)

- Removing `USE_TX_DEPENDENCIES` / the Polygon `Dependencies()` path.
- Raising the IO-bound worker pool above `NumCPU` (`EXEC3_WORKERS` companion lever — tracked separately).
- Cross-block state cache for the speculative pass.

### CODE REVIEW

**Comment by yperbasis** [comment]
Manually dispatched two workflows that are not part of this PR's automatic check list, since this PR changes parallel-execution scheduling and both suites exercise the parallel executor against engine-API block import:

- **Test Hive** — EL interop (engine API, RPC compat) with many same-sender-tx blocks: https://github.com/erigontech/erigon/actions/runs/26886064836
- **Hive EEST tests** — execution-spec-test fixtures via hive, covering the EIP-7702 authority-nonce and invalid/gapped-nonce edge cases called out in the issue's acceptance list: https://github.com/erigontech/erigon/actions/runs/26886066782

File: common/dbg/experiments.go [hunk 1]
```diff
@@ -113,6 +113,7 @@ var (
 	BatchCommitments      = EnvBool("BATCH_COMMITMENTS", true)
 	CaplinEfficientReorg  = EnvBool("CAPLIN_EFFICIENT_REORG", true)
 	UseTxDependencies     = EnvBool("USE_TX_DEPENDENCIES", false)
+	Exec3NoncePreWrites   = EnvBool("EXEC3_NONCE_PREWRITES", true)
```
Comment (taratorio) [inline common/dbg/experiments.go:116]: @yperbasis btw, 1 thing to try with this that was missed in the original issue spec is - try with `numWorkers=128` or even `numWorkers=256` for parallel exec alongside these changes

**Comment by yperbasis** [comment]
## Local `-race` validation summary

Green on this branch (darwin/arm64, go1.25.7, `EXEC3_PARALLEL=true`):
- `go test -race ./execution/stagedsync`
- `go test -race ./execution/tests` — full package, 563s
- `go test -race -p 2 ./execution/...` — green except the pre-existing crash below

**Pre-existing (not this PR): flaky `-race` crash in `execution/execmodule` / `execution/state` on macOS arm64.** Signature: `fatal error: runtime: split stack overflow` inside `racecall()` (`race_arm64.s:484`) right after `runtime.sigpanic`, with sp on a thread stack — a Go-runtime/TSAN-level crash, ~3-5s into the package, during `execmoduletester.New` while many `TestAssembleBlock*` tests run parallel executors concurrently. No `DATA RACE` report ever precedes it.

Attribution (sequential repetitions, same machine, quiet load, `EXEC3_PARALLEL=true go test -race ./execution/execmodule ./execution/state`):

| commit | crashes |
|---|---|
| `main` (ea6e9f1226) | **6/6** |
| this branch | 2/6 |

So the crash exists on `main` and is unrelated to this change (single-run A/B earlier in my investigation suggested otherwise — that was load-correlated noise; repetition settled it). The same packages pass the Linux race leg on this PR (`race-tests / tests-linux (execution-other, parallel)`). Crash logs preserved locally; happy to file a separate tracking issue for the macOS arm64 race-runtime crash if useful.

**Comment by yperbasis** [comment]
Both manually dispatched hive runs completed green:

- **Test Hive**: success, 12/12 jobs — https://github.com/erigontech/erigon/actions/runs/26886064836
- **Hive EEST tests**: success, 11/11 jobs — https://github.com/erigontech/erigon/actions/runs/26886066782

Together with the all-green PR checks (72 passing, incl. the eest parallel/race legs), this covers the EIP-7702 authority-nonce and invalid/gapped-nonce acceptance cases from the issue end-to-end.

**Comment by yperbasis** [comment]
## A/B vs main on mainnet tip (1h, parallel exec on both sides)

Both binaries from base `ea6e9f1226` (PR head `bae782ec` = base + 1 commit). Same Prysm via engine-mux. Each EL pinned to 4 P-cores. Steady-state at block 25237620, ran for **3600 s** with **n=301** both-VALID `engine_newPayloadV4` matched pairs.

| | mean | p50 | p90 | p99 |
|---|--:|--:|--:|--:|
| main `ea6e9f12` | 77.78 | 74.58 | 107.24 | 150.76 |
| PR 21605 `bae782ec` | 77.89 | 74.17 | 107.66 | 149.05 |
| Δ (main − PR), ms | −0.11 | +0.41 | −0.42 | +1.71 |

Per-block Δ: mean −0.11 ms, stdev 2.27 ms, SE ±0.13 ms → **Δ/SE = −0.82, within noise at ~2σ**. PR faster on 48.8% of blocks (coin-flip).

Mainnet has no BAL today (Glamsterdam not yet shipped), so this PR's no-BAL path is what every block hits — yet the wall-clock delta is zero. Most likely explanations: mainnet blocks have low same-sender clustering (few txs are freed by the pre-write), and/or the same-sender serialization wasn't the parallel-exec bottleneck on this workload — devnet measurements where the workload has more same-sender pressure should still show the gain.

### Correctness over the 1h window

| signal | count |
|---|--:|
| INVALID responses (either side) | **0** |
| status mismatches between main and PR | **0** |
| `panic` / `fatal` / `invalid block` / `commitment mismatch` / `nonce too high\|low` | **0** |
| process deaths | 0 |

### RAM (peak `VmRSS` across the window, sampled every 5 min)

- main: ~25–30 GB
- PR 21605: **~45 GB** — sustained ~15 GB heavier throughout the run

Run-to-run RAM variance on this rig is normally ~1–2 GB; the 15 GB delta is well outside that band. Possibly the version-map entries the PR pre-writes (one per tx, every block) accumulate more than expected in the parallel-exec working set. Worth a look before merging.

**Comment by yperbasis** [comment]
Follow-up on the macOS `-race` crash documented above: it turned out to be fully root-caused and fixable — mdbx maps landing between Go heap arenas with unmapped TSAN shadow (plus the same VA squatting causing 'too many address space collisions'). Fix in #21611 (`main` 6/6 fatal → 0/8 with the fix).

**Comment by sudeepdino008** [comment]
> Most likely explanations: mainnet blocks have low same-sender clustering (few txs are freed by the pre-write), and/or the same-sender serialization wasn't the parallel-exec bottleneck on this workload 

this contradicts what was mentioned in the ticket about the "sender-based dependency scheduling" being a large cause for the gap.
